presence/fpp/fpp/src/presence_detector.rs

```diff
@@ -0,0 +1,128 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{HashMap, VecDeque};
+use std::time::Instant;
+
+use itertools::Itertools;
+
+use crate::fspl_converter::compute_distance_meters_at_high_tx_power;
+use crate::fused_presence_utils::{
+    BleScanResult, MaybeTxPower, MeasurementConfidence, PresenceDataSource, ProximityEstimate,
+    ProximityState, DEFAULT_CONSECUTIVE_SCANS_REQUIRED,
+    DEFAULT_LONG_RANGE_DISTANCE_THRESHOLD_METERS, DEFAULT_REACH_DISTANCE_THRESHOLD_METERS,
+    DEFAULT_SHORT_RANGE_DISTANCE_THRESHOLD_METERS, DEFAULT_TAP_DISTANCE_THRESHOLD_METERS,
+};
+
+const MAX_RSSI_FILTER_VALUE: i32 = 10;
+const DEFAULT_ESTIMATED_DISTANCE_DATA_TTL_MILLIS: u128 = 4000;
+
+/// Static function for getting proximity state from threshold
+fn get_proximity_state_from_threshold(distance_meters: f64) -> ProximityState {
+    if distance_meters <= DEFAULT_TAP_DISTANCE_THRESHOLD_METERS {
+        return ProximityState::Tap;
+    }
+    if distance_meters <= DEFAULT_REACH_DISTANCE_THRESHOLD_METERS {
+        return ProximityState::Reach;
+    }
+    if distance_meters <= DEFAULT_SHORT_RANGE_DISTANCE_THRESHOLD_METERS {
+        return ProximityState::ShortRange;
+    }
+    if distance_meters <= DEFAULT_LONG_RANGE_DISTANCE_THRESHOLD_METERS {
+        return ProximityState::LongRange;
+    }
+    ProximityState::Far
+}
+
+/// Tracks and computes proximity/presence state events.
+pub struct PresenceDetector {
+    last_range_update_time: RangingUpdateTime,
+    best_proximity_estimate_per_device: HashMap<u64, ProximityEstimate>,
+    transition_history: VecDeque<ProximityState>,
+}
+
+pub struct RangingUpdateTime(u128);
+
+impl RangingUpdateTime {
+    pub fn is_expired(&self) -> bool {
+        let elapsed_real_time_millis = Instant::now().elapsed().as_millis();
+        elapsed_real_time_millis - self.0 > DEFAULT_ESTIMATED_DISTANCE_DATA_TTL_MILLIS
+    }
+
+    pub fn update(&mut self) {
+        self.0 = Instant::now().elapsed().as_millis();
+    }
+}
+
+impl PresenceDetector {
+    pub fn new() -> Self {
+        PresenceDetector {
+            last_range_update_time: RangingUpdateTime(0),
+            best_proximity_estimate_per_device: HashMap::new(),
+            transition_history: VecDeque::with_capacity(
+                (DEFAULT_CONSECUTIVE_SCANS_REQUIRED + 1).into(),
```

Comment:
What's the reason for the "+ 1" here?

Reply:
See lines 107-110 - Allocating one extra entry to accommodate the overflow, since we do push_front before truncating using consecutive scans required.